model_access_restriction/models/ir_model_access_restriction.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2025 Alberto Martínez <alberto.martinez@sygel.es>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, exceptions, fields, models
+
+
+class IrModelAccessRestriction(models.Model):
+    _name = "ir.model.access.restriction"
+    _description = "Model Access Restrictions. Not having one will disable access."
+
+    _order = "model_id,name,id"
+
+    name = fields.Char(index=True)
+    active = fields.Boolean(
+        default=True,
+    )
+    model_id = fields.Many2one(
+        string="Model",
+        comodel_name="ir.model",
+        index=True,
+        required=True,
+        ondelete="cascade",
+    )
+    groups = fields.Many2many(
+        string="Allowed Groups",
+        comodel_name="res.groups",
+        relation="model_access_restriction_group_rel",
+        column1="model_access_restriction_id",
+        column2="group_id",
+        ondelete="restrict",
+    )
+    # perm_read = fields.Boolean(string='Apply for Read', default=True)
+    # perm_write = fields.Boolean(string='Apply for Write', default=True)
+    perm_create = fields.Boolean(string="Apply for Create", default=True)
+    perm_unlink = fields.Boolean(string="Apply for Delete", default=True)
+
+    def _get_model_restrictions(self, model, operation):
+        query = """ SELECT r.id
+                    FROM ir_model_access_restriction r JOIN ir_model m ON (r.model_id=m.id)
+                    WHERE m.model=%s AND r.active AND r.perm_{operation}
+                    ORDER BY r.id
+                """.format(
+            operation=operation
+        )
+        self._cr.execute(query, (model,))
+        return self.browse(row[0] for row in self._cr.fetchall()).exists()
+
+    def _get_passed_restrictions(self, model, operation):
+        query = """ SELECT r.id
+                    FROM ir_model_access_restriction r
+                        JOIN ir_model m ON (r.model_id=m.id)
+                    WHERE m.model=%s AND r.active AND r.perm_{operation}
+                    AND (r.id IN (
+                        SELECT model_access_restriction_id
+                        FROM model_access_restriction_group_rel rg
+                        JOIN res_groups_users_rel gu ON (rg.group_id=gu.gid)
+                        WHERE gu.uid=%s)
+                    )
+                    ORDER BY r.id
+                """.format(
+            operation=operation
+        )
+        self._cr.execute(query, (model, self._uid))
+        return self.browse(row[0] for row in self._cr.fetchall()).exists()
+
+    def check_restrictions(self, model, operation):
+        if operation not in ["create", "unlink"]:
+            return True
+        model_restrictions = self._get_model_restrictions(model, operation)
+        return not model_restrictions or not (
```

Comment:
Is it worthy to have two queries?

Reply:
I think that. 

The small query is faster and will be used alone the 99% of the time

It also simplifies the development